llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -157,6 +159,55 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineSetExtract(
   return true;
 }
 
+bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorInserts(
+    MachineInstr &MI) const {
+  const Register DstReg = MI.getOperand(0).getReg();
+  unsigned DstRegLen = MRI.getType(DstReg).getElementCount().getKnownMinValue();
+  SmallVector<Register, 16> Regs;
+  MachineInstr *CurMI = &MI;
+  int NumInserts = DstRegLen;
+
+  auto IsSet = [&](const MachineInstr *MI) {
+    return MI->getOpcode() == AIE2::G_INTRINSIC &&
+           cast<GIntrinsic>(*MI).getIntrinsicID() ==
+               Intrinsic::aie2_set_I512_I128;
+  };
+
+  auto IsVInsert = [&](const MachineInstr *MI) {
+    return MI->getOpcode() == AIE2::G_INTRINSIC &&
+           cast<GIntrinsic>(*MI).getIntrinsicID() ==
+               Intrinsic::aie2_vinsert8_I512;
+  };
+
+  while (NumInserts--) {
+    const Register SrcReg = CurMI->getOperand(2).getReg();
+    const Register IdxReg = CurMI->getOperand(3).getReg();
+    const Register SclSrcReg = CurMI->getOperand(4).getReg();
+
+    auto Cst = getIConstantVRegValWithLookThrough(IdxReg, MRI);
+    if (!Cst || Cst->Value.getSExtValue() != NumInserts)
```

Comment:
This expects the lane indices to be in a specific order. Instead, we could collect them and check whether we are complete afterwards.